File: cpp/35.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <iterator>
#include <utility>
#include <map>
#include <set>
#include <string>
#include <unordered_map>
#include <climits>
#include <numeric>
using namespace std;

template<typename T>
void prints(vector<T> &v){
  cout << "[";
  for(uint i=0; i<v.size()-1; ++i)
    cout << v[i] << ", ";
  if(v.size()) cout << v[v.size()-1];
  cout << "]\n";
}

class Solution {
public:
  int searchInsert(int A[], int n, int target) {
    int l = distance(A, lower_bound(A, A+n, target));
    return l;
  }
};


int main(){
  auto sol = Solution();
  int a[] = {1,3,5,6};
  cout << sol.searchInsert(a, 4, 4);
}

File: python/35.py
class Solution(object):
    def searchInsert(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        if not nums or target < nums[0]:
            return 0

        n = len(nums)
        l, r = 0, n
        while l+1 != r:
            m = (l+r) / 2
            if nums[m] <= target:
                l = m
            else:
                r = m

        return l if l != n and nums[l] == target else l+1
